### PR TITLE
Add tracing to Trees and Deep Subtrees

### LINKED ISSUE #19: Generate Witness Data in IAVL tree by tracing operations
We'd like to add tracing for keys accessed in the IAVL tree along with the operation itself during the following operations:

- Set
- Get
- Remove

For each operation above, the keys accessed should be used to generate the existence proofs needed for performing the said operation. 

Overall, after performing a bunch of operations, an IAVL tree should have the following struct inside it to record all the operations and existence proofs required for each operation:

```
type WitnessData struct {
	Operation Operation
	Key       []byte
	Value     []byte
	Proofs    []*ics23.ExistenceProof
}
```

Also, add an option to enable/disable tracing in an IAVL tree.

### LINKED ISSUE #20: Use Witness Data generated by IAVL tree in Deep Subtrees
We'd like to be able to load Witness Data generated in #19 in a Deep Subtree and later on be able to perform all operations recorded in the Witness Data. 

When an operation is performed on the Deep Subtree, it should first check that the operation matches the current operation in the Witness Data that hasn't been performed. Also, we should verify all the existence proofs inside the Witness Data to make sure they match up with the root hash of the Deep Subtree at any point.

Parent issue: #19

### LINKED ISSUE #21: Add initial root hash in Deep Subtree
A Deep Subtree should be able to be initialized with an initial root hash since the root is set to `nil` for an empty Deep Subtree. However, if a non-nil root exists in the Deep Subtree, it should have priority over the initial root hash and the working hash of the Deep Subtree should be used as the root hash.

Add setters and getters for initial root hash along with using it in `verifyOperation` and `buildTree`.

Parent issue: #20

### LINKED ISSUE #22: Test Get operation in randomized tests
In current fuzzing tests, we only test `Set` and `Remove`. We'd also like to test the `Get` operation for deep subtrees.

### PR DESCRIPTION
**Pull request by Manav-Aggarwal**
## Overview

Closes: #19 
Closes: #20 
Closes: #21 
Closes: #22 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: deepsubtree.go [hunk 1]
```diff
@@ -186,38 +113,56 @@ func (dst *DeepSubTree) linkNode(node *Node) error {
...
 			node.rightNode, _ = dst.ndb.GetNode(node.rightHash)
-		} else if !bytes.Equal(node.rightNode.hash, node.rightHash) {
+		}
+	}
+	return nil
+}
+
+func (dst *DeepSubTree) verifyOperation(operation Operation, key []byte, value []byte) error {
+	// Verify operation is on top, look at the witness data and add the relevant existence proofs
+	if dst.witnessData != nil && dst.operationCounter < len(dst.witnessData) {
```
Comment (tzdybal) [inline deepsubtree.go]: Shouldn't you return error is `dst.operationCounter >= len(dst.witnessData)` (`else` for this `if`).

Comment (Manav-Aggarwal) [reply deepsubtree.go]: Yes, thanks for pointing this out. Addressed.

File: deepsubtree.go [hunk 2]
```diff
@@ -21,14 +20,15 @@ const (
...
+	initialRootHash  []byte        // Initial Root Hash when Deep Subtree is initialized for an already existing tree
+	witnessData      []WitnessData // Represents a trace operation along with inclusion proofs required for said operation
+	operationCounter int           // Keeps track of which operation in the witness data list the Deep Subtree is on
 }
 
 // NewDeepSubTree returns a new deep subtree with the specified cache size, datastore, and version.
 func NewDeepSubTree(db dbm.DB, cacheSize int, skipFastStorageUpgrade bool, version int64) *DeepSubTree {
 	ndb := newNodeDB(db, cacheSize, nil)
-	head := &ImmutableTree{ndb: ndb, version: version}
+	head := &ImmutableTree{ndb: ndb, version: version, skipFastStorageUpgrade: skipFastStorageUpgrade}
```
Comment (tzdybal) [inline deepsubtree.go:31]: Nice catch :+1:

File: deepsubtree.go [hunk 3]
```diff
@@ -119,10 +139,57 @@ func (dst *DeepSubTree) linkNode(node *Node) error {
...
+			"operation counter in witness data: %d should be less than length of witness data: %d",
+			dst.operationCounter,
+			len(dst.witnessData),
+		)
+	}
+	traceOp := dst.witnessData[dst.operationCounter]
+	if traceOp.Operation != operation || !bytes.Equal(traceOp.Key, key) || !bytes.Equal(traceOp.Value, value) {
+		return fmt.Errorf(
+			"traceOp in witnessData: %s, %s, %s does not match up with write operation for key: %s, value: %s",
+			traceOp.Operation, string(traceOp.Key), string(traceOp.Value), string(key), string(value),
```
Comment (tzdybal) [inline deepsubtree.go]: This method is used also for reading.
```suggestion
			"traceOp in witnessData (%s, %s, %s) does not match up with executed operation (%s, %s, %s)",
			traceOp.Operation, string(traceOp.Key), string(traceOp.Value), 
			operation, string(key), string(value),
```

File: deepsubtree_test.go [hunk 4]
```diff
@@ -369,50 +350,91 @@ func (tc *testContext) removeInDST(key []byte) error {
...
 				}
 				keys.Delete(string(keyToDelete))
+			case Get:
+				keyToGet, err := tc.getKey(true, false)
+				require.NoError(err)
+				t.Logf("%d: Get: %s\n", i, string(keyToGet))
+				err = tc.getInDST(keyToGet)
+				if err != nil {
+					t.Error(err)
+				}
```
Comment (tzdybal) [inline deepsubtree_test.go:437]: :+1:

File: mutable_tree.go [hunk 5]
```diff
@@ -126,11 +148,56 @@ func (tree *MutableTree) prepareOrphansSlice() []*Node {
 	return make([]*Node, 0, tree.Height()+3)
 }
 
-// Set sets a key in the working tree. Nil values are invalid. The given
+// Return a list of existences proofs for all keys in the given set
+func (tree *MutableTree) reapExistenceProofs(keysAccessed []string) ([]*ics23.ExistenceProof, error) {
+	existenceProofs := make([]*ics23.ExistenceProof, 0)
```
Comment (tzdybal) [inline mutable_tree.go]: Please pre-allocate the slice with capacity of `len(keysAccessed)`.

**Review by tzdybal — COMMENTED** [review]
Looks good. I left small comment regarding error message.

**Review by tzdybal — APPROVED** [review]
LGTM

**Pull request merged**